main.cpp

```diff
@@ -92,6 +92,105 @@ good to go!
  */
 
 #include <iostream>
+
+struct FloatType 
+{
+    float add(float lhs, float rhs);
+    float subtract(float lhs, float rhs );
+    float multiply(float lhs, float rhs );
+    float divide(float lhs, float rhs );
+};
+
+float FloatType::add(float lhs, float rhs)
+{
+    return lhs + rhs;
+}
+
+float FloatType::subtract(float lhs, float rhs)
+{
+    return lhs - rhs;
+}
+
+float FloatType::multiply(float lhs, float rhs)
+{
+    return lhs * rhs;
+}
+
+float FloatType::divide(float lhs, float rhs)
+{
+    if (0.0f == rhs )
+        std::cout << "\nwarning, floating point division by zero returns 'inf' !\n";
+    return lhs / rhs;
+}
+
+struct DoubleType 
+{
+    double add(double lhs, double rhs);
+    double subtract(double lhs, double rhs );
+    double multiply(double lhs, double rhs );
+    double divide(double lhs, double rhs );
+};
+
+double DoubleType::add(double lhs, double rhs)
+{
+    return lhs + rhs;
+}
+
+double DoubleType::subtract(double lhs, double rhs)
+{
+    return lhs - rhs;
+}
+
+double DoubleType::multiply(double lhs, double rhs)
+{
+    return lhs * rhs;
+}
+
+double DoubleType::divide(double lhs, double rhs)
+{
+    if (0.0 == rhs)
+        std::cout << "\nwarning, floating point division by zero returns 'inf' !\n";
+    return lhs / rhs;
+}
+
+struct IntType 
+{
+    int add(int lhs, int rhs);
+    int subtract(int lhs, int rhs );
+    int multiply(int lhs, int rhs );
+    int divide(int lhs, int rhs );
+};
+
+int IntType::add(int lhs, int rhs)
+{
+    return lhs + rhs;
+}
+
+int IntType::subtract(int lhs, int rhs)
+{
+    return lhs - rhs;
+}
+
+int IntType::multiply(int lhs, int rhs)
+{
+    return lhs * rhs;
+}
+
+int IntType::divide(int lhs, int rhs)
+{
+    if (0 == rhs)
+    {
+        std::cout << std::endl
+                  << "error, integer division by zero will crash the program!"
+                  << std::endl
+                  << "returning lhs"
+                  << std::endl;
+        return lhs;
+    }
+    else
+        return lhs / rhs;
+}
```

Comment:
2 things: 
1) If you're going to use curly braces in an `if()` block, use them in any `else` or `else if` blocks.
2) If you find yourself writing code like this:
```
int Type::someFunction()
{
    if( someExpression )
    {
        someStatement();
        return aValue;
    } 
    else 
    {
        return anotherValue;
    }
}
```
You do not need the `else` block.  Here's why:
`return` statements are the last thing that **ever happens** in a function.
When the `if()` expression is evaluated as `true` and the block is entered, the `return aValue;` is the last line of code to be executed before the function exits.
but when the `if()` expression evaluates as `false` and the block is **NOT** entered, the next thing that happens is `return anotherValue;`
you have no code after the `else` block that needs executing, since your `else` block contains a `return` statement.
that means you do **NOT** need the `else` block, and it is just causing visual clutter.
```
int Type::someFunction()
{
    if( someExpression )
    {
        someStatement();
        return aValue;
    } 
    
    return anotherValue;
}
```
- [ ] revise